tests/e2e/upgradecluster/upgradecluster_test.go

```diff
@@ -0,0 +1,460 @@
+package upgradecluster
+
+import (
+	"flag"
+	"fmt"
+	"os"
+	"strings"
+	"testing"
+
+	. "github.com/onsi/ginkgo"
+	"github.com/onsi/ginkgo/reporters"
+	. "github.com/onsi/gomega"
+	"github.com/rancher/k3s/tests/e2e"
+)
+
+// Valid nodeOS: generic/ubuntu2004, opensuse/Leap-15.3.x86_64, dweomer/microos.amd64
+var nodeOS = flag.String("nodeOS", "generic/ubuntu2004", "VM operating system")
+var serverCount = flag.Int("serverCount", 3, "number of server nodes")
+var agentCount = flag.Int("agentCount", 2, "number of agent nodes")
+
+//valid format: RELEASE_VERSION=v1.23.1+k3s2 or nil for latest commit from master
+var installType = flag.String("installType", "", "version or nil to use latest commit")
+
+func Test_E2EUpgradeValidation(t *testing.T) {
+	RegisterFailHandler(Fail)
+	flag.Parse()
+	RunSpecsWithDefaultAndCustomReporters(t, "Create Cluster Test Suite", []Reporter{
+		reporters.NewJUnitReporter("/tmp/results/junit-cc.xml"),
+	})
+}
+
+var (
+	kubeConfigFile  string
+	serverNodenames []string
+	agentNodenames  []string
+)
+
+var _ = Describe("Verify Upgrade", func() {
+	Context("Cluster :", func() {
+		It("Starts up with no issues", func() {
+			var err error
+			serverNodenames, agentNodenames, err = e2e.CreateCluster(*nodeOS, *serverCount, *agentCount, *installType)
+			Expect(err).NotTo(HaveOccurred())
+			fmt.Println("CLUSTER CONFIG")
+			fmt.Println("OS:", *nodeOS)
+			fmt.Println("Server Nodes:", serverNodenames)
+			fmt.Println("Agent Nodes:", agentNodenames)
+			kubeConfigFile, err = e2e.GenKubeConfigFile(serverNodenames[0])
+			Expect(err).NotTo(HaveOccurred())
+		})
+
+		It("Checks Node and Pod Status", func() {
+			fmt.Printf("\nFetching node status\n")
+			Eventually(func(g Gomega) {
+				nodes, err := e2e.ParseNodes(kubeConfigFile, false)
+				g.Expect(err).NotTo(HaveOccurred())
+				for _, node := range nodes {
+					g.Expect(node.Status).Should(Equal("Ready"))
+				}
+			}, "420s", "5s").Should(Succeed())
+			_, _ = e2e.ParseNodes(kubeConfigFile, true)
+
+			fmt.Printf("\nFetching Pods status\n")
+			Eventually(func(g Gomega) {
+				pods, err := e2e.ParsePods(kubeConfigFile, false)
+				g.Expect(err).NotTo(HaveOccurred())
+				for _, pod := range pods {
+					if strings.Contains(pod.Name, "helm-install") {
+						g.Expect(pod.Status).Should(Equal("Completed"), pod.Name)
+					} else {
+						g.Expect(pod.Status).Should(Equal("Running"), pod.Name)
+					}
+				}
+			}, "420s", "5s").Should(Succeed())
+			_, _ = e2e.ParsePods(kubeConfigFile, true)
+		})
+
+		It("Verifies ClusterIP Service", func() {
+			_, err := e2e.DeployWorkload("clusterip.yaml", kubeConfigFile, false)
+			if err != nil {
+				fmt.Println("Cluster IP manifest not deployed", err)
+				Expect(err).NotTo(HaveOccurred())
+			}
```

Comment:
Can be combined into one line
```suggestion
			Expect(err).NotTo(HaveOccurred(), "Cluster IP manifest not deployed")
```